### PR TITLE
Improved CI

### PR DESCRIPTION
**Pull request by jmlarson1**

### CODE REVIEW

**Comment by jared321** [comment]
Is the file `test_requirements.txt` in the pounders tests folder still needed?  If so, we might need to find a way to integrate that in a better way.

File: pounders/py/pounders.py [hunk 1]
```diff
@@ -73,11 +75,16 @@ def pounders(fun, X0, n, npmax, nfmax, gtol, delta, nfs, m, F0, xkin, L, U, prin
...
+
+        def hfun(F):
+            return np.sum(F**2)
+
         from .general_h_funs import leastsquares as combinemodels
 
     # choose your spsolver
     if spsolver == 2:
+        _FILE_LOCATION = Path(__file__).resolve().parent
+        sys.path.append(os.path.join(_FILE_LOCATION, "../../minq/py/minq5/"))
```
Comment (jared321) [inline pounders/py/pounders.py:87]: I see you imported `Path`.  Might want to use it instead of `os.path` now for the path management.  Split `"../../minq/py/minq5/"` into individual arguments to avoid using `/`.

Does this work when you use the installed python package?  Since `minq` isn't in the package, I would guess that the relative path hardcoded here will not work.

Insist that user must set the python path appropriately?  Have them set a `minq` env var that the package can use to set the python path internally?

File: .github/workflows/flake8.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      with:
+        python-version: 3.11
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install flake8
+
+    - name: Run Flake8
+      run: flake8
```
Comment (jared321) [inline .github/workflows/flake8.yml:25]: You could turn this action into a more generic "check package" action and have it call `tox -r -e check`, which includes a call to flake8.

This action would become more generic and its specification would be similar to the CI action, which might have its benefits.

Is this a temporary work around until you can figure out to get flake8 to work correctly with tox?  If so, you might want to say so in this file and what your final intentions would be once that problem is resolved.  Also, you could create an Issue and reference the issue in those comments.

File: ibcdfo_pypkg/setup.py [hunk 3]
```diff
@@ -1,47 +1,50 @@
...
-# installed with pip.  This implies that the LICENSE is in the sdist tarball for
-# those who access the code outside of pip.
-python_requires = ">=3.7"
-install_requires = ["wheel", "numpy>=1.16.5", "scipy>=1.6"]
-# TODO: Can we integrate testing using tox or other solution?
-# test_requires = ["oct2py"]
+# Changes made to python_requires should be propagated to all tox.ini and all
+# GitHub Action config files.
+python_requires = ">=3.8"
+code_requires = ["numpy", "scipy"]
```
Comment (jared321) [inline ibcdfo_pypkg/setup.py:24]: Why did you remove the version info?

File: ibcdfo_pypkg/setup.py [hunk 4]
```diff
@@ -1,47 +1,50 @@
...
 
-packages = find_packages(include=["ibcdfo", "ibcdfo.pounders"])
-# TODO: This is a hack because these are python files rather than package data,
-# which could be something like data files needed by tests.  I believe that this
-# is unnecessary if test file organization follows the pattern
-#                         test/test*.py
-package_data = {"ibcdfo.pounders": ["tests/*/*.py"]}
+package_data = {
+    "IBCDFO": [
+        "pounders/tests/regression_tests/benchmark_resutls/*.txt",
```
Comment (jared321) [inline ibcdfo_pypkg/setup.py:30]: results is misspelled.  Also, the benchmark results folder is not in the regression tests folder.

File: ibcdfo_pypkg/src/ibcdfo/load_tests.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                         python -m unittest ibcdfo
+
+    Parameters:
+        loader - the unittest.TestLoader instance doing the loading
+    """
+    here_dir = Path(__file__).resolve().parent
+    start_dir = here_dir.joinpath("tests")
+
+    print()
+    print(f"Discover tests in {start_dir}")
```
Comment (jared321) [inline ibcdfo_pypkg/src/ibcdfo/load_tests.py:33]: You can remove the printing if you want.  I included that in the template so that you can see that `coverage` uses the local clone and `nocoverage` uses the package installed in site_packages.

File: pounders/py/load_tests.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+                         python -m unittest ibcdfo.pounders
+    to run alone.
+
+    Parameters:
+        loader - the unittest.TestLoader instance doing the loading
+        suite - test suite being built at the time of call to this function
+    """
+    start_dir = Path(__file__).resolve().parent.joinpath("tests")
+
+    print(f"Discover tests in {start_dir}")
```
Comment (jared321) [inline pounders/py/load_tests.py:22]: Ditto.  No need to print.

File: ibcdfo_pypkg/tox.ini [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+#
+
+[tox]
+requires = tox>=4
+env_list = nocoverage
+
+[testenv]
+allowlist_externals = wget
+description = Run the ibcdfo full test suite
+basepython = py3
```
Comment (jared321) [inline ibcdfo_pypkg/tox.ini:13]: Does adding this make a difference?

File: ibcdfo_pypkg/tox.ini [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+description = Deploy distribution to TestPyPi for testing
+deps = twine
+commands =
+    python setup.py sdist
+    python setup.py bdist_wheel
+    # This will prompt for username/password
+    #
+    # If dist was not empty beforehand, this could potentially upload many
+    # different versions of distributions.
+    twine upload --repository testpypi dist/*
```
Comment (jared321) [inline ibcdfo_pypkg/tox.ini:54]: FYI, I never tested this section.  Note sure if you need it.

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+# Coveralls
+#   * finds the report,
+#   * shows correctly the filenames of python files inside (i.e., within the
+#     ibcdfo_pypkg directory) and outside (e.g., within subA) of the python
+#     package, and
+#   * allows users to view the coverage results for each python file inside and
+#     outside of the python package.
+# This extra care is needed as a result of developing subpackages such as subA
+# outside of the python package and including it into the package via symlink.
+#
```
Comment (jared321) [inline tox.ini:16]: Replace `subA` above with `pounders`.

File: tox.ini [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+[testenv:format]
+description = Clean-up code with black and reporting flake8 issues
+deps =
+    black
+    flake8
+skip_install = true
+commands =
+    black --config=.black {toxinidir}
+    flake8
```
Comment (jared321) [inline tox.ini:84]: I don't know if I would include flake8 in `format`.  I understand `format` as a verb and to mean "CAREFUL: this will change your files!"

As a developer, I would want to be able to run flake8 in one command, which won't alter my files, without running black, which will.

**Review by jared321 — APPROVED** [review]
Looks good to me.  Please see inline comments and questions.

Any sense in updating the year in the LICENSE file in this PR?

File: pounders/py/pounders.py (hunk 1 above)
Comment (jmlarson1) [reply pounders/py/pounders.py]: I agree. We now try the import and if the module isn't found, we point the user in the right location.

File: ibcdfo_pypkg/setup.py (hunk 4 above)
Comment (jmlarson1) [reply ibcdfo_pypkg/setup.py:30]: This has been fixed

File: ibcdfo_pypkg/setup.py (hunk 3 above)
Comment (jmlarson1) [reply ibcdfo_pypkg/setup.py:24]: Restored!

File: ibcdfo_pypkg/src/ibcdfo/load_tests.py (hunk 5 above)
Comment (jmlarson1) [reply ibcdfo_pypkg/src/ibcdfo/load_tests.py:33]: Removed!

File: pounders/py/load_tests.py (hunk 6 above)
Comment (jmlarson1) [reply pounders/py/load_tests.py:22]: resolved

File: ibcdfo_pypkg/tox.ini (hunk 7 above)
Comment (jmlarson1) [reply ibcdfo_pypkg/tox.ini:13]: If you are asking about the `wget`, then yes, it's needed. Without it, running `tox -r -e coverage` errors: 
  `coverage: failed with wget is not allowed, use allowlist_externals to allow it`

File: ibcdfo_pypkg/tox.ini (hunk 8 above)
Comment (jmlarson1) [reply ibcdfo_pypkg/tox.ini:54]: I have removed it for now.

File: tox.ini (hunk 9 above)
Comment (jmlarson1) [reply tox.ini:16]: Good catch. Resolved.

File: tox.ini (hunk 10 above)
Comment (jmlarson1) [reply tox.ini:87]: I agree that this isn't best. I've raised an issue for us to discuss how to resolve this. (#100) both and the `IBCDFO` level, and at the package level.
